primitives.lisp

```diff
@@ -1154,27 +1160,55 @@ When non-NIL the group is restored even if it already exists. This arg should
 be set to the dump filename to use for forced restore. Defaults to NIL
 
 @item class
-The window's class must match @var{class}.
+The windows class must match @var{class}.
+
+@item class-not
+The windows class must not match @var{class-not}
 
 @item instance
-The window's instance/resource name must match @var{instance}.
+The windows instance/resource name must match @var{instance}.
+
+@item instance-not
+The windows instance/resource name must not match @var{instance-not}.
 
 @item type
-The window's type must match @var{type}.
+The windows type must match @var{type}.
+
+@item type-not
+The windows type must not match @var{type-not}.
 
 @item role
-The window's role must match @var{role}.
+The windows role must match @var{role}.
+
+@item role-not
+The windows role must not match @var{role-not}.
 
 @item title
-The window's title must match @var{title}.
+The windows title must match @var{title}.
+
+@item title-not
+The windows title must not match @var{title-not}.
+
+@item match-properties-and-function
+A function that, if provided, must return true alongside the provided properties 
+in order for the rule to match. This function takes one argument, the window. 
+
+@item match-properties-or-function
+A function that, if provided and returning true, will cause the rule to match
+regardless of whether the window properties match. Takes one argument, the window.
 @end table"
   (let ((x (gensym "X")))
     `(dolist (,x ',frame-rules)
        ;; verify the correct structure
-       (destructuring-bind (frame-number raise lock
-                                         &rest keys
-                                         &key from-group create restore class instance type role title) ,x
-         (declare (ignore from-group create restore class instance type role title))
+       (destructuring-bind (frame-number raise lock &rest keys
```

Comment:
I vote to remove the `&key` parameters from the code. They are used as a way to document what keywords a frame-rule has but that is already documented in the docstring.

If we decide against it then please leave the `&rest` parameter on its own line. There is a method to the madness :grimacing:

Reply:
That works for me, I'll remove it unless theres a reason to keep them.